### PR TITLE
Integrate Grid Picker with Storybook

### PR DESCRIPTION
**Pull request by tjmw**
## What does this change?

Builds on top of #44 by integrating with Grid, allowing the marketing user to select an image from Grid instead of specifying the image manually. As with the other configurable elements, the marketing user would then transfer these to Braze as key/value pairs when configuring an in-app message.

https://trello.com/c/4QSxdwus/994-integrate-grid-picker-with-storybook

Outstanding items, which could be tackled in future PRs:

* Hosted storybooks are still on Chromatic, where the grid picker won't load because of the content security policy. They need to be moved to a gutools.co.uk domain.
* We aren't being smart about which crop size to use, we simply selected the last one on the list. We've talked about extending this to generate a `i.guim.co.uk` URL from the `media.guim.co.uk` url returned by Grid. If we did that I'd think we'd use the master crop and then apply our own quality and size params to the i.guim url.
* Known issue where the currently selected image is reset when the browser window is resized below a certain width.
* Running storybook locally with `yarn storybook` opens on localhost:6016 where the grid picker won't work, again because of the CSP issue mentioned above. You have to visit https://braze-storybook.local.dev-gutools.co.uk/ manually.
* You must have already auth'd with Grid to be able to use the picker in storybook.

## Images

Screenshot 2020-11-20 at 16 44 33

### CODE REVIEW

File: src/utils/images.ts
Comment (paulbrown1982) [inline src/utils/images.ts:4]: We should raise a ticket with Ed Tools to get object listing removed from this bucket, especially as we have now made the bucket name public...

File: .storybook/grid/GridPanel.tsx
Comment (paulbrown1982) [inline .storybook/grid/GridPanel.tsx:35]: Won't this code break if the image has no crops (which is somewhat common)?

File: .storybook/grid/GridPanel.tsx
Comment (paulbrown1982) [inline .storybook/grid/GridPanel.tsx:50]: I can't see how the inner scope `useEffect` will know the name of the `isModelOpen` variable, if it's only passed in as an array rather than an object in its second argument...

File: .storybook/grid/GridPanel.tsx
Comment (adamnfish) [inline .storybook/grid/GridPanel.tsx:63]: As a reader it isn't crystal clear to me that this "stops listening". This appears to be adding a handler that does nothing, rather than removing the existing handler? If this works then I guess this is actually mutating the existing handler reference.

No harm if this works, but if this is confusing to others as well perhaps we could think about removing the events rather than changing them to NoOps in this situation?

File: src/utils/images.ts
Comment (adamnfish) [reply src/utils/images.ts:4]: Great shout. We should also check that it's ok to make this bucket name public at all. We might wanna nuke the history here if not. My suspicion is that this bucket name is "private", in the three-tier "public", "private", "secret" categorisation, which means it shouldn't be in a public repo.

We're racking up use cases that need secrets, I wanna chat to you all about adding a backend to this where we can put them.

File: .storybook/grid/GridPanel.tsx
Comment (adamnfish) [inline .storybook/grid/GridPanel.tsx:12]: I feel like these configuration constants are a bit scattered at the moment. Not urgent for now, but we might consider colocating these definitions in the future so that:

* it's easy to see what has been configured in one place
* it's simpler to keep these configurations in sync
* the around these bits of config (specifically how we detect dev mode) is not duplicated

File: .storybook/grid/GridPanel.tsx
Comment (adamnfish) [reply .storybook/grid/GridPanel.tsx:35]: The `isValidMessage` function checks for a crop, but doesn't seem to be used. Once you have a crop, it *should* have assets available, but maybe we want to double check with EdTools that this is true - a crop will always have at least one asset. If not, we can add this check to `isValidMessage` (once that validation is used).

File: .storybook/grid/GridPanel.tsx
Comment (adamnfish) [inline .storybook/grid/GridPanel.tsx:15]: Is this used?

**Review by adamnfish — COMMENTED** [review]
Great!

This looks basically fine, but there are a couple things to look at. I do think we might want to look into the bucket string before this gets merged because life gets more difficult to fix the further that spreads. Otherwise, my other points are things we can talk about later or now, up to you both.

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx:15]: Ah great spot, thanks, I'll remove. No it's not used any longer, `@guardian/grid-client` is now doing this for us. (I'm surprised the editor/build steps didn't shout about this, perhaps there's a linting/typescript config value we're missing to make sure it does)

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx:12]: Agreed with all this 👍  As a first pass a single method to check whether we're in dev mode seems like a good place to start.

File: src/utils/images.ts
Comment (tjmw) [reply src/utils/images.ts:4]: Ahh thanks for pointing this out, that concern didn't occur to me here. Let's chat about what needs to happen with this.

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx:50]: Hey good question. The scoping here is that the `isModelOpen` const is defined outside the first argument to `useEffect` (the callback) so it closes over it. The second (optional) argument to `useEffect` is the list of dependencies, so at the end of the current render cycle the effect will be called only if `isModalOpen` has changed.

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx:35]: We've replaced our own homegrown `isValidMessage` with `!postMessageService.isValid` from `@guardian/grid-client` which, if I understand correctly, will return `false` if there are no crops. I'll double check that though, it'd be good to be sure. I'd say if the validation _isn't_ right it'd be good to capture those rules in `guardian/grid-client` rather than expanding on them here.

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx:63]: Oh weirdly I'd seen examples where the return value of the `.on` function was a function to de-register the handler, but looking at the docs that's not the case. So I must have been mistaken. I'll fix. I think the more standard `api.off` is the way to go.

File: .storybook/grid/GridPanel.tsx
Comment (tjmw) [reply .storybook/grid/GridPanel.tsx]: I've done a little refactoring around this. I'm sure we can improve things further but I think I might save that for a future PR.

**Comment by tjmw** [comment]
Hi @adamnfish @paulbrown1982 thanks for all the feedback, we've been through and I think we've addressed everything at this point. Regarding the allow-listing of the bucket - we decided that since we only need that for development that rather than being too specific we can loosen the validation for dev only and allow any https://s3-eu-west-1.amazonaws.com/ URL (without specifying the bucket). I think that's enough to unblock this PR - what do you reckon?

**Review by paulbrown1982 — APPROVED** [review]
👍  LGTM

**Review by adamnfish — APPROVED** [review]

**Pull request merged**
